app/models/audit.rb

```diff
@@ -0,0 +1,99 @@
+require 'syslog'
+require 'logger'
+require 'time'
+
+module Audit
+  class << self
+    def notice msg, msgid, data
+      logger.info LogMessage.new msg, msgid, data, Syslog::LOG_NOTICE
+    end
+
+    def logger
+      @logger ||= Rails.logger
+    end
+
+    attr_writer :logger
+  end
+
+  module SDID
+    CONJUR_PEN = 43838
+    def self.conjur_sdid label
+      [label, CONJUR_PEN].join('@').freeze
+    end
+
+    POLICY = conjur_sdid 'policy'
+    AUTH = conjur_sdid 'auth'
+    SUBJECT = conjur_sdid 'subject'
+    ACTION = conjur_sdid 'action'
+  end
+
+  class LogMessage < String
+    def initialize msg, msgid, structured_data = nil, severity = nil
+      super msg
+      @msgid = msgid
+      @structured_data = structured_data
+      @severity = severity
+    end
+
+    attr_reader :msgid, :structured_data, :severity
+  end
+
+  # Middleware to store request ID in a thread variable
+  class RememberUuid < Struct.new :app
+    def call env
+      Thread.current[:request_id] = env["action_dispatch.request_id"]
+      app.call env
+    end
+  end
+
+  class RFC5424Formatter
+    SEVERITY_MAP = {
+      Logger::Severity::DEBUG => Syslog::LOG_DEBUG,
+      Logger::Severity::ERROR => Syslog::LOG_ERR,
+      Logger::Severity::FATAL => Syslog::LOG_CRIT,
+      Logger::Severity::INFO => Syslog::LOG_INFO,
+      Logger::Severity::WARN => Syslog::LOG_WARNING
+    }
+
+    # TODO
+    FACILITY = 32
+
+    def call severity, time, progname, msg
+      severity = if msg.respond_to? :severity
+        msg.severity
+      else
+        SEVERITY_MAP[Logger::Severity.const_get severity] rescue severity
+      end
+      sd = msg.structured_data if msg.respond_to? :structured_data
+      timestamp = time.utc.iso8601 3
+
+      # Will be filled in by syslogd, however we might want to use a
+      # different configured name on appliance instead.
+      hostname = nil
+
+      progname ||= 'conjur'
+
+      # Use the request UUID when called with REST. Otherwise use PID.
+      pid = Thread.current[:request_id] || Process.pid
+
+      msgid = msg.msgid if msg.respond_to? :msgid
+      sd = format_sd sd
+
+      fields = [timestamp, hostname, progname, pid, msgid, sd, msg]
+      ["<#{severity + FACILITY}>1", *fields.map {|x| x || '-'}].join(" ") + "\n"
```

Comment:
Is there any possibility of a newline in `x`?

Reply:
Depends on the input. Formatters don't usually guard against it, so I chose not to here, too.